### ISSUE TITLE
imagePullSecrets on a Deployment not been propagated to pods

### ISSUE THREAD

**Issue by bitgandtter** [kubernetes/kubernetes#57801]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
I have a Deployment configuration like:

```
- apiVersion: extensions/v1beta1
  kind: Deployment
  metadata:
    annotations:
      kompose.cmd: kompose convert -o k8s
      kompose.version: 1.6.0 (e4adfef)
    creationTimestamp: null
    labels:
      io.kompose.service: user
    name: user
  spec:
    replicas: 1
    strategy: {}
    template:
      metadata:
        creationTimestamp: null
        labels:
          io.kompose.service: user
      spec:
        containers:
        - env:
          - name: APP_DEBUG
            value: "0"
          - name: APP_ENV
            value: prod
          image: "user/private-repo"
          name: user
          resources: {}
        restartPolicy: Always
        imagePullSecrets:
          - name: regsecret
  status: {}
```
where **regsecret** is a secret created following [the official doc](https://kubernetes.io/docs/concepts/containers/images/#specifying-imagepullsecrets-on-a-pod) but the created pod failed to pull the image because authentication and the pod does not mount the specified secret, see describe:

```
Name:           user-798fc86589-2lmd4
Namespace:      default
Node:           ip-172-31-9-21/172.31.9.21
Start Time:     Wed, 03 Jan 2018 19:13:17 +0000
Labels:         io.kompose.service=service
                pod-template-hash=3549742145
Annotations:    <none>
Status:         Pending
IP:             10.1.98.124
Controlled By:  ReplicaSet/user-798fc86589
Containers:
  service:
    Container ID:   
    Image:          user/private-repo
    Image ID:       
    Port:           <none>
    State:          Waiting
      Reason:       ErrImagePull
    Ready:          False
    Restart Count:  0
    Environment:
      APP_DEBUG:  0
      APP_ENV:    prod
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-4tdzg (ro)
Conditions:
  Type           Status
  Initialized    True 
  Ready          False 
  PodScheduled   True 
Volumes:
  default-token-4tdzg:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-4tdzg
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason                 Age                From                     Message
  ----     ------                 ----               ----                     -------
  Normal   Scheduled              11m                default-scheduler        Successfully assigned service-798fc86589-2lmd4 to ip-172-31-9-21
  Normal   SuccessfulMountVolume  11m                kubelet, ip-172-31-9-21  MountVolume.SetUp succeeded for volume "default-token-4tdzg"
  Warning  Failed                 10m (x2 over 11m)  kubelet, ip-172-31-9-21  Failed to pull image "user/private-repo": rpc error: code = Unknown desc = Error response from daemon: repository primoshome/service not found: does not exist or no pull access
  Warning  Failed                 10m (x2 over 11m)  kubelet, ip-172-31-9-21  Error: ErrImagePull
  Normal   BackOff                10m (x5 over 11m)  kubelet, ip-172-31-9-21  Back-off pulling image "user/private-repo"
  Normal   SandboxChanged         10m (x7 over 11m)  kubelet, ip-172-31-9-21  Pod sandbox changed, it will be killed and re-created.
  Normal   Pulling                10m (x3 over 11m)  kubelet, ip-172-31-9-21  pulling image "user/private-repo"
  Warning  Failed                 1m (x52 over 11m)  kubelet, ip-172-31-9-21  Error: ImagePullBackOff
```

**What you expected to happen**:

i expect the pod to be configured with the secret and be able to pull the image from the private repository

**How to reproduce it (as minimally and precisely as possible)**:

exec above configuration with any private repository

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.9.0
- Cloud provider or hardware configuration: aws
- OS (e.g. from /etc/os-release): ubuntu
- Kernel (e.g. `uname -a`): Linux ip-172-31-10-97 4.4.0-64-generic #85-Ubuntu SMP Mon Feb 20 11:50:30 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: ubuntu official kubernetes (juju)
- Others:

**Comment by liggitt** [comment]
pull secrets don't seem to be included in `describe` output. can you verify the pull secret is included in `kubectl get pod user-798fc86589-2lmd4 -o yaml`?

**Comment by zhangwei0181** [comment]
I suffered the same problem today.

**Comment by zhangwei0181** [comment]
Below configurations exists in the pod spec. But it is not work.
 imagePullSecrets:
  - name: registrysecret

**Comment by nkwangleiGIT** [comment]
@bitgandtter  @zhangwei0181 what's the version of Docker? and you can also check the data of secret registrysecret to confirm it's in correct format like blow:

```
{"192.168.1.123":{"Username":"test","Password":"test","Email":"test@tenxcloud.com"}}
```

As we tried on k8s 1.6.11, imagePullSecrets should work normally.

**Comment by bitgandtter** [comment]
@nkwangleiGIT docker version 17.12 also notice that k8s is on version 1.9.0. Also yes the data is that format

**Comment by nkwangleiGIT** [comment]
@bitgandtter ok, I'll have a try on 1.9.0 later to see if it'll have this issue , will update here later, thanks.

**Comment by wolfroma** [comment]
It seems I have the same problem
```
imagePullSecrets:
      - name: regsecret
```
It appears in yaml output but the describe doesn't show it.


```
Name:           frs-79969778c9-pxn5s
Namespace:      my-ns
Node:           node1/10.21.140.194
Start Time:     Thu, 04 Jan 2018 09:17:11 -0800
Labels:         app=frs
                pod-template-hash=3552533475
Annotations:    <none>
Status:         Pending
IP:             10.233.102.143
Controlled By:  ReplicaSet/frs-79969778c9
Containers:
  frs:
    Container ID:
    Image:          masked-docker.jfrog.io/frs:testingkube
    Image ID:
    Port:           8443/TCP
    State:          Waiting
      Reason:       ImagePullBackOff
    Ready:          False
    Restart Count:  0
    Liveness:       http-get http://:8443/masked/healthy delay=300s timeout=1s period=10s #success=1 #failure=3
    Environment:    <none>
    Mounts:
      /etc/ssl/certs from certs (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-scgzn (ro)
Conditions:
  Type           Status
  Initialized    True
  Ready          False
  PodScheduled   True
Volumes:
  certs:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  docker-certificates-frs
    Optional:    false
  default-token-scgzn:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-scgzn
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     <none>
Events:
  Type     Reason                 Age                            From               Message
  ----     ------                 ----                           ----               -------
  Normal   Scheduled              <invalid>                      default-scheduler  Successfully assigned frs-79969778c9-pxn5s to node1
  Normal   SuccessfulMountVolume  <invalid>                      kubelet, node1     MountVolume.SetUp succeeded for volume "certs"
  Normal   SuccessfulMountVolume  <invalid>                      kubelet, node1     MountVolume.SetUp succeeded for volume "default-token-scgzn"
  Normal   SandboxChanged         <invalid>                      kubelet, node1     Pod sandbox changed, it will be killed and re-created.
  Normal   BackOff                <invalid> (x3 over <invalid>)  kubelet, node1     Back-off pulling image "masked-docker.jfrog.io/frs:testingkube"
  Warning  Failed                 <invalid> (x3 over <invalid>)  kubelet, node1     Error: ImagePullBackOff
  Normal   Pulling                <invalid> (x2 over <invalid>)  kubelet, node1     pulling image "masked-docker.jfrog.io/frs:testingkube"
  Warning  Failed                 <invalid> (x2 over <invalid>)  kubelet, node1     Failed to pull image "masked-docker.jfrog.io/frs:testingkube": rpc error: code = Unknown desc = Error response from daemon: Get https://masked-docker.jfrog.io/v2/frs/manifests/testingkube: unknown: Authentication is required
  Warning  Failed                 <invalid> (x2 over <invalid>)  kubelet, node1     Error: ErrImagePull`

```
I have slightly different format of registry though

```
{
  "auths": {
    "https://masked-docker.jfrog.io": {
      "username": "masked",
      "password": "masked",
      "email": "masked@mail.com",
      "auth": "base64token"
    }
  }
}

```


Environment:
Kubernetes version (use kubectl version): v1.9.0 , server: v1.9.0+coreos.0
Cloud provider or hardware configuration: vm
OS (e.g. from /etc/os-release): centos 7
Kernel (e.g. uname -a): Linux node1 3.10.0-514.10.2.el7.x86_64 #1 SMP Fri Mar 3 00:04:05 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
Install tools: kubespray
Others:
   docker version: 17.09.1-ce

**Comment by liggitt** [comment]
describe is just a human readable version of the pod. it does not include all fields. `get -o yaml` is authoritative.

you are likely encountering https://github.com/kubernetes/kubernetes/issues/57427, fixed in 1.9.1

**Comment by liggitt** [comment]
closing, fixed in https://github.com/kubernetes/kubernetes/releases/tag/v1.9.1

**Comment by beatlejuse** [comment]
I have the same problem.
In DaemonSet it works, but in deployment - no.
I add imagePullSecrets to deployment and to ServiceAccount both.
I see imagePullSecrets string in "kubectl edit po" but pod stay in status "ImagePullBackOff"

Environment:
Bare-metal CentOS 7.5
Docker 18.06.1-ce
Hyperkube 1.11.2
etcd 3.3.9
kubectl 1.11.2

**Comment by beatlejuse** [comment]
its my manifests:
https://gist.github.com/beatlejuse/3bd6875b574fc2940a282366217b1686
https://gist.github.com/beatlejuse/7afe3be88cd3896c398db38f3c5983cc
and deploy script:
```
kubectl -n kube-system apply -f $WORKDIR/kube-dns-cr.yaml
  sleep 1
kubectl -n kube-system patch serviceaccount kube-dns -p '{"imagePullSecrets": [{"name": "registry"}]}'
kubectl -n kube-system patch serviceaccount default -p '{"imagePullSecrets": [{"name": "registry"}]}'

envsubst < $WORKDIR/kube-dns.yaml | kubectl create -f -
```

**Comment by datapresso** [comment]
Still see this issue in k8 v1.11.1 when we tried to bring in the k8-dashboard (https://github.com/kubernetes/dashboard) in our private registry.

**Comment by StefanSchubert** [comment]
Seems to be still a problem. k8s 1.12 and imagePullSecret does not work and also is not known by deployments apiVersion: apps/v1 

So private registry => kills deployment as structure? really?

**Comment by Scavallarin** [comment]
Same problem for me, tried so many ways but i can not pull from a docker private repository. It looks like that "imagePullSecrets:" in the .yml is not even considered. 
Any one has find any turn around? 
Thanks

**Comment by beatlejuse** [comment]
@Scavallarin i fix it.
show your manifest.

**Comment by kbielak** [comment]
same here

minikube version: `v0.30.0`
docker: `18.09.ce`
kubectl:
```
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.2", GitCommit:"17c77c7898218073f14c8d573582e8d2313dc740", GitTreeState:"clean", BuildDate:"2018-10-30T21:39:16Z", GoVersion:"go1.11.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-26T16:44:10Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```
MacOS Mojave 0.14.1

looking for a workaround, anyone?

**Comment by jwenz723** [comment]
same here. Running on EKS v1.10.11-eks.

**Comment by zakkg3** [comment]
In my case was a indentation problem:

Image section should be placed in container specification. ImagePullSecret should be placed in spec section so proper yaml file looks like this (please note indent):

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: test-pod-deployment
  labels:
    app: test-pod
spec:
  replicas: 3
  selector:
    matchLabels:
      app: test-pod
  template:
    metadata:
      labels:
        app: test-pod
    spec:
      containers:
       - name: test-pod
         image: <private-registry>
      imagePullSecrets:
       - name: regcred
```

**Comment by geosword** [comment]
For me, the deployment I was creating was specifying its own specific service account. I had assumed that the default service account would be attached in addition to any explicitly defined service accounts in the pod / daemonset / deployment manifest.
Once I added the docker secret to the service account mentioned in the deployment, it worked as expected.

**Comment by wgv-ebishop** [comment]
@geosword what is the process for adding "the docker secret to the service account"?

**Comment by kesterriley** [comment]
Instead of adding the secret in to the yaml file as detailed above, I resolved this by copying the results of my docker login, in to the following locations:


sudo mkdir -p /root/.docker/
sudo cat <<EOF >  /root/.docker/config.json
{
	"auths": {
		"my-docker-repo.com:5000": {
			"auth": "my encrypted password generated from docker login"
		}
	},
	"HttpHeaders": {
		"User-Agent": "Docker-Client/18.06.1-ce (linux)"
	}
}
EOF
sudo mkdir -p /var/lib/kubelet/
sudo cp /root/.docker/config.json /var/lib/kubelet/config.json


I hope this helps someone.

**Comment by donbowman** [comment]
For this having this issue... you need to make imagePullSecrets a peer to container, e.g. not under the container. Shift it left 2 in your yaml.

just for replication controllers (eg. Deployment, Statefulset).

**Comment by sylhero** [comment]
anything after
 ```
imagePullSecrets:
       - name: regcred
``` 
has errors for example:
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: frontend
spec:
  replicas: 1
  strategy:
    type: RollingUpdate
  template:
    metadata:
      labels:
        app: frontend
    spec:
      containers:
      - name: frontend
        image: images.ormuco.com:5000/ormuco/spiderkube/frontend:latest
        ports:
        - containerPort: 80
        readinessProbe:
          httpGet:
            path: /
            port: 80
          initialDelaySeconds: 30
          periodSeconds: 3
        resources:
          limits:
            cpu: "0.1"
            memory: 100Mi

**Comment by JanBielawski** [comment]
Still not work!

Even if you use "apiVersion: extensions/apps/v1" as the K8s V1.13 documentation recoments.

Get 
`error validating data: ValidationError(Deployment.spec.template.spec.containers[0]): unknown field "imagePullSecrets" in io.k8s.api.core.v1.Container; if you choose to ignore these errors, turn validation off with --validate=false`

For:
```
  spec:
    containers:
      - name: ada
         image: kpserviceplatform.azurecr.io/activedirectoryadapter:2.0.0.0
         ports:
         - containerPort: 80
         imagePullSecrets:
         - name: kpserviceplatform.azurecr.io
```

**Comment by donbowman** [comment]
in your example you still have it in the wrong spot. it needs to be peer to containers. not under a specific container.

**Comment by bertoost** [comment]
facing the same issue.. I have added the docker-registry secret to the right namespace, values are correct, but it looks like the `Deployment` is not reading it. I have the `imagePullSecrets` on the same level as `containers` (within the `spec`). But no luck.

Kubernetes dashboard shows this error message;
```
Failed to pull image "***/my-image:latest": rpc error: code = Unknown desc = Error response from daemon: pull access denied for ***/my-image, repository does not exist or may require 'docker login'
```

**Comment by kesterriley** [comment]
I fixed this by adding the file created by docker login to the home directory
________________________________

**Comment by bertoost** [comment]
Yeah I tried that too.. but that also doesn't seem to work for me

**Comment by andreas-wolf** [comment]
I had the same problem and besides I had the wrong indent for imagePullSecrets the next problem was that the docs were a bit misleading. They have examples like that:

```
{
    "auths": {
        "https://index.docker.io/v1/": {
            "auth": "c3R...zE2"
        }
    }
}
```
but actually if you have your docker image prefixed with something like privateRepository:5000/imageName and you use the example

> kubectl create secret docker-registry regcred --docker-server=<your-registry-server> --docker-username=<your-name> --docker-password=<your-pword> --docker-email=<your-email>

make sure that  &lt;your-registry-server&gt; equals privateRepository:5000 and **not** something like https://privateRepository/

I mean now it's pretty obvious, but I've overlooked that error for about two days.

**Comment by bertoost** [comment]
@andreas-wolf that makes sense, but I have configured my registry behind Traefik (proxy) on a `registry.mydomain.com` .. so I also used that in creating the secrets. I also can connect from external devices, but now I have to the same from the kubernetes deployments too

**Comment by bertoost** [comment]
Damn! I found it... 
My private registry secret had the wrong type;

`type: kubernetes.io/dockercfg`
 &nbsp; &nbsp; vs 
`type: kubernetes.io/dockerconfigjson` <-- right!

**Comment by jthera** [comment]
I think there's a subset of folks experience this who are running a private repo with self-signed certs.    Deploying a replicaset to 3 nodes including the master (it's a test cluster in a private LAN, no judging!) all the deployments to the master are fine, all the deployments to remote nodes are failing with "certificate signed by unknown authority"    

Unlikely that this is a bug - more likely just a gap in documentation for this edge case.

**Comment by rlig** [comment]
Gentleman, just add secret to appropriate namespace which is supposed to use it.

for example:
![image](https://user-images.githubusercontent.com/32437712/60021348-26b07180-9692-11e9-997b-559bb8430672.png)

**Comment by ThibaultJanBeyer** [comment]
My issue was that I had a wrong format of the secret:

It has to be:
```
{"auths":{"https://test.com/v2/":{"username": … … …
```

But I only had `{"auths":{"test.com":{"username": … … …`

So make sure to have the `https://` and `/v2/` part.

**Comment by willmao** [comment]
we are facing a similar issue, one deployment can not pull a image while others are all fine, clone the deployment with a different name and the new deployment can pull image successfully, do not what is wrong

**Comment by nshah14** [comment]
You need to check if you are deploying in similar namespace as creds are created... imagepullsecrets works just fine but its in align with containers not as to one containers mapped variable.

**Comment by kovkev** [comment]
Hi kubernetes! I'm also trying to set something like this up and there is no clear way to troubleshoot the issue.

I use https://github.com/bazelbuild/rules_k8s#aliasing-eg-k8s_deploy where I specify the namespace to be "default".

I set in my deployment.yaml file a `imagePullSecrets: - name: regcred` under the `spec.template.spec` level. 

I ran `KUBECONFIG=/etc/kubernetes/admin.conf kubectl patch serviceaccount default -p '{"imagePullSecrets": [{"name": "regcred"}]}'`

I see for `kubectl describe pod my-first-ever-pod` : `Error response from daemon: pull access denied for human/forum, repository does not exist or may require 'docker login': denied: requested access to the resource is denied`.

I can do `docker login https://mywebsite.com/` and I get `Login Succeeded` without having to put in my username:password

I made sure to put `"HTTPS://mywebsite.com:5000/V2/"` (in lowercase) in the `auth` section in the docker config file before I generated the `regcred` secret.

I am able to run `docker pull mywebsite.com:5000/some/repository:dev` and see all the layers being downloaded.

I see `sudo curl --user testuser:testpassword --cacert /usr/local/share/ca-certificates/mywebsite.registry.com/ca.crt  -X GET https://mywebsite.com:5000/v2/_catalog`
\> `{"repositories":["a/repository"]}`

and

`sudo curl --user testuser:testpassword --cacert /usr/local/share/ca-certificates/mywebsite.registry.com/ca.crt  -X GET https://mywebsite.com:5000/v2/human/forum/tags/list`
\> `{"name":"a/repository","tags":["dev"]}`

====

I am wondering: I would like to see what is the

```
kubectl describe pod my-first-pod...
...
Volumes:
  default-token-mtz9g:
    Type:        Secret
```

, where it is located and what it contains? Does it actually contain the right authentication?

Also, I would like to inspect the logs of the kubernetes API. Where could I inspect these? Do I have to run `kubeadm init` with some specific parameters to turn on logging? because I don't see anything in `/var/log/kube-apiserver.log`.

**Comment by cdelacruzpinto** [comment]
I'm having the following problem:

- I Have a pull docker token from Gitlab (they allow to create accounts that allow only pulling from the registry of a single project), which expired, and two namespaces with same container image deployed. Both deployments in both namespaces pull using the same secret.
- In one of the namespaces, I replaced the secret and deleted the deployment. Created it again. Pull works.
- Then in the other namespace, I replaced also the secret. Killed the pods. Still the deployment (which in this namespace was not replaced) is unable to pull the containers.
- The base64 for the docker secret is exactly equal in both namespaces. imagePullSecret is set to the proper secret name in both of them. Is there some known bug regarding secrets not being refreshed after replacing them ?
